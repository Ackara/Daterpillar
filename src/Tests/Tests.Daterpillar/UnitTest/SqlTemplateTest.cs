using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Gigobyte.Daterpillar.Transformation;
using Gigobyte.Daterpillar.Transformation.Template;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Runtime.CompilerServices;
using Telerik.JustMock;
using Telerik.JustMock.Helpers;

namespace Tests.Daterpillar.UnitTest
{
    [TestClass]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
    public class SqlTemplateTest
    {
        [ClassCleanup]
        public static void Cleanup()
        {
            ApprovalTests.Maintenance.ApprovalMaintenance.CleanUpAbandonedFiles();
        }

        /// <summary>
        /// Assert <see cref="SqlTemplate.Transform(Schema)"/> returns a well formatted T-SQL schema
        /// when all settings are enbaled.
        /// </summary>
        [TestMethod]
        [Owner(Dev.Ackara)]
        public void GenerateSqlSchemaWithSettingsEnabled()
        {
            // Arrange
            var settings = new SqlTemplateSettings()
            {
                AddScript = true,
                CommentsEnabled = true,
                DropDatabaseIfExist = true,
            };

            RunTemplateTest(settings);
        }

        /// <summary>
        /// Assert <see cref="SqlTemplate.Transform(Schema)"/> returns a well formatted T-SQL schema
        /// when all settings are disabled.
        /// </summary>
        [TestMethod]
        [Owner(Dev.Ackara)]
        public void GenerateSqlSchemaWithSettingsDisabled()
        {
            // Arrange
            var settings = new SqlTemplateSettings()
            {
                AddScript = false,
                CommentsEnabled = false,
                DropDatabaseIfExist = false,
            };

            RunTemplateTest(settings);
        }

        /// <summary>
        /// Assert <see cref="SqlTemplate.Transform(Schema)"/> assign a value to <see
        /// cref="Index.Table"/> if null or empty.
        /// </summary>
        [TestMethod]
        [Owner(Dev.Ackara)]
        public void AutoFillSqlIndexTableNameIfMissing()
        {
            // Arrange
            var sut = new SqlTemplate(new SqlTemplateSettings(), Mock.Create<ITypeNameResolver>());
            var schema = SampleData.CreateSchema();
            schema.Tables[0].Indexes[0].Table = "";

            // Act
            var script = sut.Transform(schema);

            // Assert
            Approvals.Verify(script);
        }

        private void RunTemplateTest(SqlTemplateSettings settings, [CallerMemberName]string name = null)
        {
            // Arrange
            var schema = SampleData.CreateSchema(name);
            var managerTable = SampleData.CreateTableSchema("Manager");
            managerTable.Comment = "this is a comment";
            schema.Tables.Add(managerTable);

            var mockTypeResolver = Mock.Create<ITypeNameResolver>();
            mockTypeResolver.Arrange(x => x.GetName(Arg.IsAny<DataType>()))
                .Returns("int")
                .OccursAtLeast(1);

            var sut = new SqlTemplate(settings, mockTypeResolver);

            // Act
            var tsql = sut.Transform(schema);

            // Assert
            mockTypeResolver.AssertAll();
            Approvals.Verify(tsql);
        }
    }
}