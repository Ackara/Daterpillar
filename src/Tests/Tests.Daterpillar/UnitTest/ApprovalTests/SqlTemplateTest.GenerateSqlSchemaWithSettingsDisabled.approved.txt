CREATE DATABASE [GenerateSqlSchemaWithSettingsDisabled];

CREATE TABLE [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Employee]
(
	[Id] int NOT NULL IDENTITY(1, 1),
	[Full_Name] int not null default 'n/a',
	[Salary] int not null,
	CONSTRAINT [fkey1] FOREIGN KEY ([Id]) REFERENCES [Card]([Id])
);

ALTER TABLE [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Employee] ADD CONSTRAINT [GenerateSqlSchemaWithSettingsDisabled_Employee_pk] PRIMARY KEY ([Id] ASC);
CREATE INDEX [generatesqlschemawithsettingsdisabled_employee_idx1] ON [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Employee] ([Id] ASC);
CREATE TABLE [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Manager]
(
	[Id] int NOT NULL IDENTITY(1, 1),
	[Full_Name] int not null default 'n/a',
	[Salary] int not null,
	CONSTRAINT [fkey1] FOREIGN KEY ([Id]) REFERENCES [Card]([Id])
);

ALTER TABLE [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Manager] ADD CONSTRAINT [GenerateSqlSchemaWithSettingsDisabled_Manager_pk] PRIMARY KEY ([Id] ASC);
CREATE INDEX [generatesqlschemawithsettingsdisabled_manager_idx2] ON [GenerateSqlSchemaWithSettingsDisabled].[dbo].[Manager] ([Id] ASC);
